Docs: Improve Argument Clinic tutorial preamble (#107323)
python/cpython: Doc/howto/clinic.rst

Tutorial
========

The best way to get a sense of how Argument Clinic works is to
convert a function to work with it.  Here, then, are the bare
minimum steps you'd need to follow to convert a function to
work with Argument Clinic.  Note that for code you plan to
check in to CPython, you really should take the conversion farther,
using some of the :ref:`advanced concepts <clinic-howtos>`
you'll see later on in the document,
like :ref:`clinic-howto-return-converters`
and :ref:`clinic-howto-self-converter`.
But we'll keep it simple for this walkthrough so you can learn.

First, make sure you're working with a freshly updated checkout

> of the CPython trunk.

Author: I don't think this brings much value in the tutorial. The tutorial is targeted at the core dev and contributor audience; working on an old checkout seems like a very little likely scenario, IMO.

Reviewer: While they know, it might be useful as a reminder.  If being a few commits behind is not an issue, then the reminder is not essential and can be removed.

Author: I think the tutorial currently mixes two things:

1. using a specific example (pickle.Pickler.dump)
2. trying to teach general guidelines for how to apply clinic to CPython (if x, then don't do y, if z then you can do w, etc.)

I'd like the tutorial to be repeatable and to not drift into digression after digression. The sentence in question could be a part of a "How to apply Argument Clinic to CPython" recipe, instead of part of the tutorial.

Author: I think it'll be hard to align the tutorial according to Diátaxis principles until we've extracted the "how to apply Argument Clinic to CPython" bits from it. Perhaps we should do that first.

What do you think, @ezio-melotti?

> In this tutorial, you will convert a C function from having custom
> :ref:`argument parsing code <arg-parsing>`
> to using Argument Clinic to generate that boilerplate code.

Reviewer: I feel this sentence could be better, maybe something like:
```suggestion
In this tutorial, you will learn how to use Argument Clinic to generate
:ref:`argument parsing code <arg-parsing>` for a function that
was using custom code.
```

The current sentence uses the imperative form to tell you what to do, without telling you why, in a way that sounds a bit harsh.  "Converting a function" seems something an howto would cover, whereas "learn how to use AC (to convert functions)" seems a better fit for the tutorial.

In addition, using "we" ("*we will learn how*") sounds "friendlier" to me, even thought technically only the reader is learning. This might require a refactoring of the whole tutorial, even though I see there are some instances of "we" later.  I don't remember if diataxis has any preference regarding this.

> If it contains any of the following format units...:

Author: We should consider removing this from the tutorial, in my opinion. Alternatives and information are tutorial anti-patterns.

Reviewer: You could state what format units are relevant for the tutorial, and link to the advanced converters section for the others.  IOW state what works, rather than what doesn't work.
